### PR TITLE
[mui-datatables] use correct type for customBodyRender tableMeta.tableData

### PR DESCRIPTION
**Pull request by smonist**
As of now, when using a customBodyRender, there is a tableMeta object returned with a property named tableData. This property turns out to be an array of the tables' data. The typings indicate a different structure. This PR fixes this issue.

This behaviour can be observed when supplying a [customBodyRender](https://github.com/gregnb/mui-datatables#column-options) for a column that makes use of [tableMeta.tableData](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/mui-datatables/index.d.ts#L73). When actually logging tableMeta.tableData the output is just a plain array of the tables' rows. There are no data and index properties.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [ ] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/gregnb/mui-datatables/blob/master/src/MUIDataTable.js#L1127
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/mui-datatables/index.d.ts#L73

### CODE REVIEW

File: types/mui-datatables/index.d.ts [hunk 1]
```diff
@@ -70,7 +70,7 @@ export interface MUIDataTableMeta {
     columnIndex: number;
     rowData: any[];
     rowIndex: number;
-    tableData: MUIDataTableData[];
+    tableData: any[];
```
Comment (favna) [inline types/mui-datatables/index.d.ts:73]:
```suggestion
    tableData: T[];
```

And change the interface to 
```ts
export interface MUIDataTableMeta<T = any>
```

That'll allow better type resolutions for the `tableData`

**Review by favna — CHANGES_REQUESTED** [review]

File: types/mui-datatables/index.d.ts (hunk 1 above)
Comment (smonist) [reply types/mui-datatables/index.d.ts]: thank you for your suggestion! i just pushed the changes :)

**Review by favna — APPROVED** [review]

**Review by andrewbranch — CHANGES_REQUESTED** [review]
This is one where a test that shows how this feature can be used would be _really_ helpful.

**Comment by smonist** [comment]
> This is one where a test that shows how this feature can be used would be _really_ helpful.

The main cause of the issues I am having with these types stems from the mui-datatables library itself. When accessing tableMeta.tableData, the data returned sometimes is an array of objects, sometimes an array of strings. The library maintainer said they do not want to fix this and instead added another option that should be used.

Related issues:
[https://github.com/gregnb/mui-datatables/pull/933](https://github.com/gregnb/mui-datatables/pull/933)
[[#1750](https://github.com/gregnb/mui-datatables/issues/1750)](https://github.com/gregnb/mui-datatables/issues/1750)

Since this happens at runtime, I found no easy way to showcase this behaviour in a test that relies solely on types.

I created a [codesandbox](https://codesandbox.io/p/sandbox/immutable-darkness-34no2u) with a minimal setup that shows the issue with mui-datatables. The lower datatable works as expected while the upper returns wrong results. To reproduce:

- open codesandbox, open browser console
- you will see console.logs of arrays of arrays
- type something in the mui-datatable search
- console.log now logs arrays of objects, despite being called from the same function

TLDR: mui-datatables has inconsistencies, it will no be fixed. The proposed solution offers typings that cover both (bugged and non-bugged) scenarios.

**Review by andrewbranch — APPROVED** [review]
Yikes. Thanks for taking the time to explain. I feel like there’s probably a more precise way to represent this, but the fact that the runtime behavior is unreliable makes me feel like it doesn’t matter much.

**Comment by smonist** [comment]
Ready to merge
